tests_e2e/tests/publish_hostname/publish_hostname.py

```diff
@@ -0,0 +1,200 @@
+        retries = 3
+
+        while retries > 0 and not status_updated:
+            agent_status_time = self._context.vm.get_instance_view().vm_agent.statuses[0].time
+            if agent_status_time != last_agent_status_time:
+                status_updated = True
+                log.info("Agent reported status at {0}".format(last_agent_status_time))
+            else:
+                retries -= 1
+                sleep(60)
+
+        if not status_updated:
+            fail("Agent hasn't reported status since {0} and ssh connection failed. Use the serial console in portal "
+                 "to check the contents of '/sys/class/net/eth0/operstate'. If the contents of this file are 'up', "
+                 "no further action is needed. If contents are 'down', that indicates the network interface is down "
+                 "and more debugging needs to be done to confirm this is not caused by the agent.\n VM: {1}\n RG: {2}"
+                 "\nSubscriptionId: {3}\nUsername: {4}\nPassword: {5}".format(last_agent_status_time,
+                                                                              self._context.vm,
+                                                                              self._context.vm.resource_group,
+                                                                              self._context.vm.subscription,
+                                                                              self._context.username,
+                                                                              self._vm_password))
+
+    def retry_ssh_if_connection_reset(self, command: str, use_sudo=False):
+        # The agent may bring the network down and back up to publish the hostname, which can reset the ssh connection.
+        # Adding retry here for connection reset.
+        retries = 3
+        while retries > 0:
+            try:
+                return self._ssh_client.run_command(command, use_sudo=use_sudo)
+            except CommandError as e:
+                retries -= 1
+                retryable = e.exit_code == 255 and "Connection reset by peer" in e.stderr
+                if not retryable or retries == 0:
+                    raise
+                log.warning("The SSH operation failed, retrying in 30 secs")
+                sleep(30)
+
+    def run(self):
+        # Add password to VM and log. This allows us to debug with serial console if necessary
+        self.add_vm_password()
+
+        # This test looks up what hostname is published to dns. Check that the tools necessary to get hostname are
+        # installed, and if not install them.
+        lookup_cmd, dns_regex = self.check_and_install_dns_tools()
+
+        # Enable agent hostname monitoring
+        log.info("Executing script update-waagent-conf to enable agent hostname monitoring")
+        result = self._ssh_client.run_command("update-waagent-conf Provisioning.MonitorHostName=y "
```

Comment:
we shouldn't enable Hostname monitoring on distros where it is disabled (e.g. Ubuntu)

Reply:
why shouldn't we test hostname path for those distros? It's possible that user can enable them at any time. That's what recent github issue reported that hostname changes not publishing to dns

Reply:
we should test them, but we should not enable this setting on distros where it is disabled